src/Processors/Transforms/ColumnGathererTransform.h

```diff
@@ -134,14 +137,17 @@ void ColumnGathererStream::gather(Column & column_res)
     if (next_required_source == -1)
     {
         /// Start new column.
-        cur_block_preferred_size = std::min(static_cast<size_t>(row_sources_end - row_source_pos), block_preferred_size);
-        column_res.reserve(cur_block_preferred_size);
+        /// Actually reserve works only for fixed size columns.
+        /// So it's safe to ignore preferred size in bytes and call reserve for number of rows.
+        size_t size_to_reserve = std::min(static_cast<size_t>(row_sources_end - row_source_pos), block_preferred_size_rows);
+        column_res.reserve(size_to_reserve);
     }
 
-    size_t cur_size = column_res.size();
     next_required_source = -1;
 
-    while (row_source_pos < row_sources_end && cur_size < cur_block_preferred_size)
+    while (row_source_pos < row_sources_end
+        && column_res.size() < block_preferred_size_rows
+        && column_res.allocatedBytes() < block_preferred_size_bytes)
```

Comment:
really minor. maybe it makes sense to use `byteSize()` because I'd expect `block_size_bytes` to refer to the actual data size.

Reply:
I checked that `allocatedBytes` is [used](https://github.com/ClickHouse/ClickHouse/blob/5a024dfc0936e062770d0cfaad0805b57c1fba17/src/Processors/Merges/Algorithms/MergedData.h#L125) in`MergedData` and did the same. But we can change it in both places.